### PR TITLE
Add video filter capability in preview screen

### PR DESCRIPTION
**Pull request by nainkunal933**
**Issue #1253:**

**Description of changes: Updated `startVideoPreviewForVideoInput` method to support filters in the preview window. Updated the browser demo to showcase the functionality and updated the UI.**

**Testing**

1. Have you successfully run `npm run build:release` locally? Yes
2. How did you test these changes? Demo app and unit tests
3. Can these changes be tested using the demo application? If yes, which demo application can be used to test it? Yes, use the application under demos/browser
4. Do you add, modify, or delete public API definitions? If yes, has that been reviewed and approved? No
5. Do you change the wire protocol, e.g. the request method? If yes, has that been reviewed and approved? No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/devicecontroller/DefaultDeviceController.test.ts [hunk 1]
```diff
@@ -2398,28 +2398,102 @@ describe('DefaultDeviceController', () => {
...
+      expect(element.srcObject === (await gUM.returnValues.pop()));
+      expect(spy1.notCalled).to.be.true;
+      expect(spy2.calledTwice).to.be.true;
+      expect(gUM.calledOnce).to.be.true;
+      spy1.restore();
+      spy2.restore();
+      gUM.restore();
+    });
+
+    it('connects the active device video stream to the video element if there is no transform device stream', async () => {
```
Comment (ltrung) [inline test/devicecontroller/DefaultDeviceController.test.ts:2427]: Can we just keep the original unit tests name `connects the video stream to the video element` and create a new one for transform device? Or say regular device or something.

File: test/devicecontroller/DefaultDeviceController.test.ts [hunk 2]
```diff
@@ -2398,28 +2398,102 @@ describe('DefaultDeviceController', () => {
...
+      deviceController.startVideoPreviewForVideoInput(element);
+      const spy1 = sinon.spy(deviceController, 'releaseMediaStream');
+      const spy2 = sinon.spy(DefaultVideoTile, 'disconnectVideoStreamFromVideoElement');
+      const gUM = sinon.spy(navigator.mediaDevices, 'getUserMedia');
+      await deviceController.chooseVideoInputDevice(stringDeviceId);
+      deviceController.startVideoPreviewForVideoInput(element);
+
       expect(spy1.called).to.be.false;
       expect(spy2.called).to.be.false;
-      expect(gUM.calledOnce).to.be.false;
```
Comment (ltrung) [inline test/devicecontroller/DefaultDeviceController.test.ts:2422]: What is the assertion here for gUM? Should only be called once right? Don't think we need to call `chooseVideoInputDevice` a second time.

File: test/devicecontroller/DefaultDeviceController.test.ts [hunk 3]
```diff
@@ -2398,28 +2398,102 @@ describe('DefaultDeviceController', () => {
...
+      const spy1 = sinon.spy(deviceController, 'releaseMediaStream');
+      const spy2 = sinon.spy(DefaultVideoTile, 'disconnectVideoStreamFromVideoElement');
+      const gUM = sinon.spy(navigator.mediaDevices, 'getUserMedia');
+      await deviceController.chooseVideoInputDevice(newStringDeviceId);
       deviceController.startVideoPreviewForVideoInput(element);
+
+      expect(spy1.calledOnce).to.be.true;
+      expect(spy2.calledOnce).to.be.true;
+      expect(gUM.calledOnce).to.be.true;
+      spy1.restore();
```
Comment (ltrung) [inline test/devicecontroller/DefaultDeviceController.test.ts:2481]: Could we do a before and after for preview and move all these spy and gUm setup and clean up there?

File: test/devicecontroller/DefaultDeviceController.test.ts [hunk 4]
```diff
@@ -2398,28 +2398,102 @@ describe('DefaultDeviceController', () => {
...
+      spy2.restore();
+      gUM.restore();
+    });
+
+    it('disconnects and releases the video stream if the video element stream is different from the active video stream', async () => {
+      await deviceController.chooseVideoInputDevice(stringDeviceId);
+      deviceController.startVideoPreviewForVideoInput(element);
+      const newStringDeviceId = 'device-id-with-transformation';
+      const spy1 = sinon.spy(deviceController, 'releaseMediaStream');
+      const spy2 = sinon.spy(DefaultVideoTile, 'disconnectVideoStreamFromVideoElement');
```
Comment (ltrung) [inline test/devicecontroller/DefaultDeviceController.test.ts:2473]: Do we need to check for `connectVideoStreamToVideoElement` also?

File: test/devicecontroller/DefaultDeviceController.test.ts [hunk 5]
```diff
@@ -2398,28 +2398,102 @@ describe('DefaultDeviceController', () => {
...
+      expect(spy1.notCalled).to.be.true;
       expect(spy2.calledOnce).to.be.true;
       expect(gUM.calledOnce).to.be.true;
       spy1.restore();
       spy2.restore();
       gUM.restore();
     });
 
-    it('does not disconnect or connect the video stream for preview if no active video exists', () => {
+    it('reuses the original video stream to create a transform device stream', async () => {
```
Comment (ltrung) [inline test/devicecontroller/DefaultDeviceController.test.ts:2445]: What is the different between this one and the one above `connects the transform video stream to the video element if there is a transform device stream` ?

Comment (ltrung) [inline test/devicecontroller/DefaultDeviceController.test.ts:2445]: Need a case with multiple video frame processor

File: demos/browser/app/meetingV2/meetingV2.ts [hunk 6]
```diff
@@ -2862,6 +2941,14 @@ export class DemoMeetingApp
       return new SegmentationProcessor();
     }
 
+    if (videoFilter === 'Resize (1/1)') {
+      return new ResizeProcessor(1);  // 1/1 Aspect Ratio
+    }
+
+    if (videoFilter === 'Resize (9/16)') {
```
Comment (ltrung) [inline demos/browser/app/meetingV2/meetingV2.ts:2948]: Can we just do one? Not sure if we need both.

File: demos/browser/app/meetingV2/meetingV2.ts [hunk 7]
```diff
@@ -2242,6 +2256,29 @@ export class DemoMeetingApp
     }
   }
 
+  populateFilterList(
```
Comment (ltrung) [inline demos/browser/app/meetingV2/meetingV2.ts:2259]: Can't we just hard-code it in the html page?

File: demos/browser/app/meetingV2/meetingV2.ts [hunk 8]
```diff
@@ -811,8 +825,8 @@ export class DemoMeetingApp
           this.audioVideo.stopVideoPreviewForVideoInput(
             document.getElementById('video-preview') as HTMLVideoElement
           );
+          await this.discardVideoFilterStream();
```
Comment (ltrung) [inline demos/browser/app/meetingV2/meetingV2.ts:828]: Can you put a comment here why we need to have this order?

File: demos/browser/app/meetingV2/meetingV2.ts [hunk 9]
```diff
@@ -746,6 +748,18 @@ export class DemoMeetingApp
       );
     });
 
+    if(!this.defaultBrowserBehaviour.supportsCanvasCapturedStreamPlayback())  {
+      document.getElementById('video-input-filter-container').className += ' no-canvas-playback';
```
Comment (ltrung) [inline demos/browser/app/meetingV2/meetingV2.ts:752]: Can we just set the style to none like other places in the code? Also do you need the check for unifiedplan like the other place in the code too?
```
if (
      this.defaultBrowserBehaviour.supportsCanvasCapturedStreamPlayback() &&
      this.enableUnifiedPlanForChromiumBasedBrowsers
    )
```

File: demos/browser/app/meetingV2/meetingV2.ts [hunk 10]
```diff
@@ -2325,6 +2362,48 @@ export class DemoMeetingApp
     }
   }
 
+  private async populateVideoPreviewFilterInputList(): Promise<void> {
```
Comment (ltrung) [inline demos/browser/app/meetingV2/meetingV2.ts:2365]: Can't we reuse `populateVideoFilterInputList`?

**Comment by ltrung** [comment]
Please rename the commit to mention this is an API change. E.g., `Add support for video processing in startVideoPreviewForVideoInput`

File: CHANGELOG.md [hunk 11]
```diff
@@ -11,10 +11,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Add `audioInputMuteStateChanged` to the `DeviceChangeObserver` interface. This is called whenever the device is changed or is muted or unmuted, allowing applications to adapt to OS-level mute state for input devices.
 - Added Android WebView Sample UI test to workflow.
 - Add a new optional API `getVideoTileForAttendeeId` in `VideoTileController` and raise the `tileWillBePausedByDownlinkPolicy` event for empty video tiles.
+- Add documentation in `Video Processing APIs` on how to add filters in the preview window.
```
Comment (devalevenkatesh) [inline CHANGELOG.md:14]: nit: We backtick only actual coding constructs like Classes, Functions.

File: guides/10_Video_Processor.md [hunk 12]
```diff
@@ -42,6 +42,7 @@ The `DefaultVideoTransformDevice` uses `VideoFrameProcessorPipeline` under the h
 #### Construction and Starting Video Processing
 
 The construction of the `DefaultVideoTransformDevice` will not start the camera or start processing. The method `meetingSession.audioVideo.chooseVideoInputDevice` is needed to be called. The device controller will use the inner `Device` to acquire the source `MediaStream` and start the processing pipeline at the same frame rate.
+`Inner device` in this context refers to the original video stream coming from the selected camera.
```
Comment (devalevenkatesh) [inline guides/10_Video_Processor.md:45]: nit: You can say "Inner `Device`".

File: guides/10_Video_Processor.md [hunk 13]
```diff
@@ -161,3 +163,23 @@ await meetingSession.audioVideo.chooseVideoInputDevice(transformDevice);
 meetingSession.audioVideo.startLocalVideo();
 
 ```
+
+### Custom processor usage in the preview window
+
+```typescript
+import ResizeProcessor from './videofilter/ResizeProcessor';
```
Comment (devalevenkatesh) [inline guides/10_Video_Processor.md:170]: Builders may not understand importing from './videofilter/ResizeProcessor'. I see above in this guide there is an example of a class showing implementation of `VideoResizeProcessor`. Can we provide similar implementation example for `ResizeProcessor`?

File: test/devicecontroller/DefaultDeviceController.test.ts (hunk 4 above)
Comment (nainkunal933) [reply test/devicecontroller/DefaultDeviceController.test.ts:2473]: I think we should. I will add that expect statement in my update.

File: CHANGELOG.md (hunk 11 above)
Comment (nainkunal933) [reply CHANGELOG.md]: Updated, thank you.

File: guides/10_Video_Processor.md (hunk 12 above)
Comment (nainkunal933) [reply guides/10_Video_Processor.md]: Updated, thank you

File: README.md [hunk 14]
```diff
@@ -213,6 +213,12 @@ Tags are fetched in order to correctly generate versioning metadata.
 To view code coverage results open `coverage/index.html` in your browser
 after running `npm run test`.
 
+> If you run `npm run test` and the tests are running but the coverage report is not getting generated
```
Comment (nainkunal933) [inline README.md:216]: @michhyun1 Please review the following documentation that I added and provide your feedback. I want you to review this since this might be helpful in improving GitHub actions in the future; I don't think there is a need right now.

The unit test GitHub actions is now working after calling `await device.stop();`

File: demos/browser/app/meetingV2/meetingV2.ts (hunk 8 above)
Comment (nainkunal933) [reply demos/browser/app/meetingV2/meetingV2.ts:830]: I have added the comment. Good call. Thanks!!

File: demos/browser/app/meetingV2/meetingV2.ts (hunk 6 above)
Comment (nainkunal933) [reply demos/browser/app/meetingV2/meetingV2.ts:2950]: I think these two are good implementations. I do not feel strongly either way though. Let me know if you want me to remove; which one should I remove then?

File: guides/10_Video_Processor.md (hunk 13 above)
Comment (nainkunal933) [reply guides/10_Video_Processor.md]: I have updated the import and naming. Can you please take a look?

File: test/devicecontroller/DefaultDeviceController.test.ts (hunk 5 above)
Comment (nainkunal933) [reply test/devicecontroller/DefaultDeviceController.test.ts:2522]: Added

File: demos/browser/app/meetingV2/meetingV2.ts (hunk 10 above)
Comment (nainkunal933) [reply demos/browser/app/meetingV2/meetingV2.ts:2367]: We can't reuse `populateVideoFilterInputList` because there are minor changes in the implementation of the method. Specifically, `populateVideoPreviewFilterInputList` calls `populateFilterList` while `populateVideoFilterInputList` calls `populateInMeetingDeviceList`.

File: test/devicecontroller/DefaultDeviceController.test.ts (hunk 2 above)
Comment (nainkunal933) [reply test/devicecontroller/DefaultDeviceController.test.ts:2422]: I agree that gUM assertion is not great. I have changed the assertion from `expect(gUM.calledOnce).to.be.false;` to `expect(gUM.called).to.be.false;`. Since, we do not call `chooseVideoInputDevice()` gUM should not be called.

File: test/devicecontroller/DefaultDeviceController.test.ts (hunk 5 above)
Comment (nainkunal933) [reply test/devicecontroller/DefaultDeviceController.test.ts:2522]: There is a slight difference between the two. Reuse one is not expecting the following two expect statements. My goal with `connects the transform video stream to the video element if there is a transform device stream` is that the stream on the canvas element is the transform stream and not the original stream.

```expect(device.getInnerDevice() === stringDeviceId);
expect(element.srcObject === (await gUM.returnValues.pop()));```

File: test/devicecontroller/DefaultDeviceController.test.ts (hunk 1 above)
Comment (nainkunal933) [reply test/devicecontroller/DefaultDeviceController.test.ts:2511]: I already have one for transform device: `connects the transform video stream to the video element if there is a transform device stream`. 

I think I should rename the active device to original device. Active device is vague as both original and transform stream can be an active stream.

File: README.md (hunk 14 above)
Comment (michhyun1) [reply README.md:216]: This is a good call out - this only occurs when initializing a video processor, right? all the other tests still pass?

File: demos/browser/app/meetingV2/meetingV2.ts [hunk 15]
```diff
@@ -746,6 +748,18 @@ export class DemoMeetingApp
       );
     });
 
+    if(!this.defaultBrowserBehaviour.supportsCanvasCapturedStreamPlayback())  {
+      document.getElementById('video-input-filter-container').className += ' no-canvas-playback';
+    }
+
+    let videoInputFilter = document.getElementById('video-input-filter') as HTMLInputElement;
+    videoInputFilter.addEventListener('change', async () => {
+      this.log('Video input filter is changed');
```
Comment (michhyun1) [inline demos/browser/app/meetingV2/meetingV2.ts:757]: Do we need this? We print the selected filter two lines down also

File: demos/browser/app/meetingV2/meetingV2.ts [hunk 16]
```diff
@@ -746,6 +748,18 @@ export class DemoMeetingApp
...
+      this.selectedVideoFilterItem = <VideoFilterName>videoInputFilter.value;
```
Comment (michhyun1) [inline demos/browser/app/meetingV2/meetingV2.ts:758]:
```
selectElement.addEventListener('change', (event) => {
  const result = document.querySelector('.result');
  result.textContent = `You like ${event.target.value}`;
});
```
Can we get the value from the callback object `event`?

File: demos/browser/app/meetingV2/meetingV2.ts (hunk 8 above)
Comment (michhyun1) [reply demos/browser/app/meetingV2/meetingV2.ts:830]: I thought we already kill the mediaStream in stopVideoPreviewForVideoInput? I think we only need to stop the transform processor, right?

File: demos/browser/app/meetingV2/meetingV2.ts (hunk 10 above)
Comment (michhyun1) [reply demos/browser/app/meetingV2/meetingV2.ts:2367]: can we just call `populateInMeetingDeviceList` conditionally based on if were in a meeting or not? It'd save some code here I think

File: demos/browser/app/meetingV2/meetingV2.ts [hunk 17]
```diff
@@ -2325,6 +2364,48 @@ export class DemoMeetingApp
...
+    );
+  }
+
+  private async discardVideoFilterStream(): Promise<void>  {
+    this.log('Clearing filter variables and discarding the video stream');
+    this.chosenVideoFilter = 'None';
+    this.selectedVideoFilterItem = 'None';
+    await this.audioVideo.chooseVideoInputDevice(null);
+    this.chosenVideoTransformDevice?.stop();
+  }
```
Comment (michhyun1) [inline demos/browser/app/meetingV2/meetingV2.ts:2407]: I tihnk the man point of this is not to discard the video stream but rather to stop the video processor

File: test/devicecontroller/DefaultDeviceController.test.ts (hunk 3 above)
Comment (nainkunal933) [reply test/devicecontroller/DefaultDeviceController.test.ts]: Yes. I have added that in my latest push.

File: demos/browser/app/meetingV2/styleV2.scss [hunk 18]
```diff
@@ -480,18 +480,34 @@ svg {
   }
 
   .video-preview {
```
Comment (michhyun1) [inline demos/browser/app/meetingV2/styleV2.scss:482]: Were you able to test these on multiple devices? I'm seeing px being used which makes me think this wont be responsive. Have you tried this on small mobile devices and also widescreen monitors?

File: demos/browser/app/meetingV2/meetingV2.ts (hunk 9 above)
Comment (nainkunal933) [reply demos/browser/app/meetingV2/meetingV2.ts:752]: I do not have to check for unified plan as far as I understand. The code snippet shared is the one used for checking for Segmentation processor. In my code, I am just checking whether the browser supports canvas capture playback. This check is done because iOS browsers will not be able to support video filters as they lack canvas capture stream playback.

File: src/devicecontroller/DefaultDeviceController.ts [hunk 19]
```diff
@@ -636,14 +636,28 @@ export default class DefaultDeviceController
...
+    // The video stream should be released only if the video stream in the HTML element is different from current stream.
+    // When a filter is applied the existing stream is reused and in that case the stream should not be released.
+    if (stream && this.activeDevices['video'].stream !== stream) {
+      this.releaseMediaStream(element.srcObject as MediaStream);
+      DefaultVideoTile.disconnectVideoStreamFromVideoElement(element, false);
+    }
+
+    if (this.chosenVideoTransformDevice) {
+      DefaultVideoTile.connectVideoStreamToVideoElement(
+        this.chosenVideoTransformDevice.outputMediaStream,
```
Comment (michhyun1) [inline src/devicecontroller/DefaultDeviceController.ts:650]: What happens if you choose a filter when there is no video preview stream? How does it get the inner stream?

File: demos/browser/app/meetingV2/styleV2.scss (hunk 18 above)
Comment (nainkunal933) [reply demos/browser/app/meetingV2/styleV2.scss:482]: Yes I tested my changes on multiple screen sizes.

File: src/devicecontroller/DefaultDeviceController.ts (hunk 19 above)
Comment (nainkunal933) [reply src/devicecontroller/DefaultDeviceController.ts:650]: The code in that scenario will not call this method and it will get to the inner stream by calling `chooseInputIntrinsicDevice`. `chooseInputIntrinsicDevice` is called by `selectVideoFilterByName` and that is connected to UI element's event handler. This is at a high level and I skipped some calls in this explanation.

File: demos/browser/app/meetingV2/meetingV2.ts (hunk 8 above)
Comment (nainkunal933) [reply demos/browser/app/meetingV2/meetingV2.ts:830]: Yes I see your point. I will not try to remove the stream again as that is just useless processing. Thank you.

File: demos/browser/app/meetingV2/meetingV2.ts (hunk 10 above)
Comment (nainkunal933) [reply demos/browser/app/meetingV2/meetingV2.ts:2367]: Not exactly. `populateInMeetingDeviceList` is used for populating buttons while `populateFilterList` is used for populating options. Also, `populateInMeetingDeviceList` has additional logic for adding additional toggles which is usually "Blue" and "SMTPE Color Bars". There is no such additional toggle logic in populateFilterList so the method params are different. Also, it doesn't make sense to call `populateInMeetingDeviceList` for populating filters nomenclature wise.

The methods have similarity but they are different and there can be some reuse but that will come at the cost of complex if else blocks with minimal improvements.

File: demos/browser/app/meetingV2/meetingV2.ts [hunk 20]
```diff
@@ -2325,6 +2363,47 @@ export class DemoMeetingApp
...
+
+    this.populateFilterList(
+      'video-input-filter',
+      genericName,
+      filters
+    );
+  }
+
+  private async stopVideoProcessor(): Promise<void>  {
+    this.log('Clearing filter variables and discarding the video stream');
```
Comment (michhyun1) [inline demos/browser/app/meetingV2/meetingV2.ts:2401]: This comment isnt exactly accurate - we are stopping the transform device

File: demos/browser/app/meetingV2/meetingV2.ts (hunk 6 above)
Comment (michhyun1) [reply demos/browser/app/meetingV2/meetingV2.ts:2948]: I think its okay to stick with one - after all, this is a demo. We just need to show how its done once.

File: test/devicecontroller/DefaultDeviceController.test.ts [hunk 21]
```diff
@@ -1948,6 +1949,12 @@ describe('DefaultDeviceController', () => {
   });
 
   describe('releaseMediaStream', () => {
+    it('returns immidiately if the stream is null', async () => {
```
Comment (michhyun1) [inline test/devicecontroller/DefaultDeviceController.test.ts:1952]: nit typo: immediately

File: demos/browser/app/meetingV2/meetingV2.ts (hunk 20 above)
Comment (nainkunal933) [reply demos/browser/app/meetingV2/meetingV2.ts:2401]: Thanks, updated pushing in next commit

File: demos/browser/app/meetingV2/meetingV2.ts (hunk 6 above)
Comment (nainkunal933) [reply demos/browser/app/meetingV2/meetingV2.ts:2948]: Thanks @michhyun1, I will remove the 1/1 aspect ratio resize processor and keep the 9/16.

File: README.md (hunk 14 above)
Comment (ltrung) [inline README.md:216]: Is there a reason to use quote here? Don't think we use it before.

File: README.md [hunk 22]
```diff
@@ -213,6 +213,12 @@ Tags are fetched in order to correctly generate versioning metadata.
...
+> then you might have a resource clean up issue. In Mocha v4.0.0 or newer the implementation was changed so 
+> that the Mocha processes will not force exit when the test run is complete. <br /> For example, if you have a 
+> `DefaultVideoTransformDevice` in your unit test then you must call `await device.stop();` to clean up the
+> resources and not run into this issue. You can also look into the usage of `done();` in the Mocha documentation.
```
Comment (ltrung) [inline README.md:220]: Can we provide the link to the documentation?

File: demos/browser/app/meetingV2/meetingV2.ts [hunk 23]
```diff
@@ -161,9 +162,9 @@ const VOICE_FOCUS_SPEC = {
 
 const MAX_VOICE_FOCUS_COMPLEXITY: VoiceFocusModelComplexity | undefined = undefined;
 
-type VideoFilterName = 'Emojify' | 'CircularCut' | 'NoOp' | 'Segmentation' | 'None';
+type VideoFilterName = 'Emojify' | 'CircularCut' | 'NoOp' | 'Segmentation' | 'Resize (9/16)' | 'None';
```
Comment (ltrung) [inline demos/browser/app/meetingV2/meetingV2.ts:165]: I think people often write aspect ratio as 16:9 right? Also can we use 1:1 or something that does not match with any screensize?

Comment (nainkunal933) [reply demos/browser/app/meetingV2/meetingV2.ts:165]: Yes 16/9 is what is called landscape mode and it is the most common aspect ratio. It is also the default aspect ratio in our meeting demo (almost exact). 

9/16 is a different aspect ratio (most commonly used for mobile phones). 9/16 is better than 1/1 as the change in video stream is significant and it helps showcase the functionality better.

File: README.md (hunk 22 above)
Comment (nainkunal933) [reply README.md:220]: Sure, I have added the link to relevant Mocha documentation.

File: demos/browser/app/meetingV2/meetingV2.ts (hunk 23 above)
Comment (ltrung) [reply demos/browser/app/meetingV2/meetingV2.ts:165]: yeah but that means it kinda weird if I test it in a mobile phone right? I just think we should choose something that is not popular so no matter what device we test it will change.

File: demos/browser/app/meetingV2/meetingV2.ts [hunk 24]
```diff
@@ -746,6 +748,17 @@ export class DemoMeetingApp
       );
     });
 
+    if(!this.defaultBrowserBehaviour.supportsCanvasCapturedStreamPlayback())  {
```
Comment (ltrung) [inline demos/browser/app/meetingV2/meetingV2.ts:751]: Do we need to check for unified plan here like in other places?

File: demos/browser/app/meetingV2/meetingV2.ts (hunk 10 above)
Comment (ltrung) [reply demos/browser/app/meetingV2/meetingV2.ts:2366]: Blue and SMPTPE Color Bars is the video input source and not filter.
I think we can do better here by refactoring these methods.

File: demos/browser/app/meetingV2/meetingV2.ts [hunk 25]
```diff
@@ -2242,6 +2257,29 @@ export class DemoMeetingApp
     }
   }
 
+  populateFilterList(
```
Comment (ltrung) [inline demos/browser/app/meetingV2/meetingV2.ts:2260]: nit: rename to preview specific such as `populateVideoPreviewFilterList`

File: guides/10_Video_Processor.md [hunk 26]
```diff
@@ -142,7 +143,8 @@ async process(buffers: VideoFrameBuffer[]): Promise<VideoFrameBuffer[]> {
 }
 ```
 
-Usage of the custom processor looks like the following:
+## Video Processing Usage
+### Custom processor usage in the meeting window
```
Comment (ltrung) [inline guides/10_Video_Processor.md:147]: meeting and preview window are only for our demo so we probably don't want to use it here. Maybe change the word to `during meeting` and `during video preview`.

File: test/devicecontroller/DefaultDeviceController.test.ts [hunk 27]
```diff
@@ -2408,48 +2415,182 @@ describe('DefaultDeviceController', () => {
 
   describe('preview', () => {
     let element: HTMLVideoElement;
+    let spy1: sinon.SinonSpy;
```
Comment (ltrung) [inline test/devicecontroller/DefaultDeviceController.test.ts:2418]: can we actually name spy1,2,3 to something specific like gUM?

File: demos/browser/app/meetingV2/meetingV2.ts (hunk 24 above)
Comment (nainkunal933) [reply demos/browser/app/meetingV2/meetingV2.ts:751]: I added a reply to this question up above as well:

"I do not have to check for unified plan as far as I understand. The code snippet shared is the one used for checking for Segmentation processor. In my code, I am just checking whether the browser supports canvas capture playback. This check is done because iOS browsers will not be able to support video filters as they lack canvas capture stream playback."

File: README.md (hunk 14 above)
Comment (nainkunal933) [reply README.md:216]: I used a quote here to bring attention to the unexpected behavior. Let me know if you want me to change it.

File: demos/browser/app/meetingV2/meetingV2.ts (hunk 23 above)
Comment (nainkunal933) [reply demos/browser/app/meetingV2/meetingV2.ts:165]: In the mobile phone, the default aspect ratio is 16/9 currently so adding 9/16 filter is a noticeable change.

File: test/devicecontroller/DefaultDeviceController.test.ts [hunk 28]
```diff
@@ -2408,48 +2415,182 @@ describe('DefaultDeviceController', () => {
...
+
+      expect(spy1.called).to.be.false;
+      expect(spy2.called).to.be.false;
+      expect(spy3.called).to.be.false;
+      expect(gUM.called).to.be.false;
+    });
+
+    it('connects the transform video stream to the video element if there is a transform device stream', async () => {
+      // It is important that the domMockBehavior setup is before the original video stream generation to mock browser behavior.
+      domMockBehavior.createElementCaptureStream = new MediaStream();
```
Comment (ltrung) [inline test/devicecontroller/DefaultDeviceController.test.ts:2450]: nit: Can we create the stream add the video track before setting it to domMockBehavior.

File: guides/10_Video_Processor.md (hunk 26 above)
Comment (nainkunal933) [reply guides/10_Video_Processor.md:147]: Sounds good, thank you.

File: test/devicecontroller/DefaultDeviceController.test.ts [hunk 29]
```diff
@@ -2408,48 +2415,182 @@ describe('DefaultDeviceController', () => {
...
+      domMockBehavior.createElementCaptureStream.addTrack(new MediaStreamTrack());
+      audioVideoController = new NoOpAudioVideoController();
+      const processor = new TestVideoFrameProcessor1();
+      const firstTransformDevice = new DefaultVideoTransformDevice(logger, stringDeviceId, [
+        processor,
+      ]);
+      await deviceController.chooseVideoInputDevice(firstTransformDevice);
+      deviceController.startVideoPreviewForVideoInput(element);
+      const firstMediaStream = await gUM.returnValues[0];
+      // In the unit test, there is no UI so no call to openVideoInputFromSelection is made.
```
Comment (ltrung) [inline test/devicecontroller/DefaultDeviceController.test.ts:2502]: This is not related to UI. In browser, calling stop stream track will make the stream to inactive. our dom mock builder does not implement this behavior.

File: demos/browser/app/meetingV2/meetingV2.ts (hunk 25 above)
Comment (nainkunal933) [reply demos/browser/app/meetingV2/meetingV2.ts:2260]: Done

File: test/devicecontroller/DefaultDeviceController.test.ts [hunk 30]
```diff
@@ -2408,48 +2415,182 @@ describe('DefaultDeviceController', () => {
...
+      await deviceController.chooseVideoInputDevice(newStringDeviceId);
+      deviceController.startVideoPreviewForVideoInput(element);
+
+      expect(spy1.calledOnce).to.be.true;
+      expect(spy2.calledOnce).to.be.true;
+      expect(spy3.calledTwice).to.be.true;
+      expect(gUM.calledTwice).to.be.true;
+    });
+
+    it('does not disconnect and release the video stream if the video element stream is same as the active video stream', async () => {
```
Comment (ltrung) [inline test/devicecontroller/DefaultDeviceController.test.ts:2574]: Great work on the unit tests.

File: README.md (hunk 14 above)
Comment (ltrung) [reply README.md:216]: Quote means you copy it from somewhere not to highlighting though.

File: demos/browser/app/meetingV2/meetingV2.ts (hunk 24 above)
Comment (ltrung) [reply demos/browser/app/meetingV2/meetingV2.ts:751]: We check for enableUnifiedPlanForChromiumBasedBrowsers in other filter population. 
```
    if (
      this.defaultBrowserBehaviour.supportsCanvasCapturedStreamPlayback() &&
      this.enableUnifiedPlanForChromiumBasedBrowsers
    ) {
```

File: test/devicecontroller/DefaultDeviceController.test.ts (hunk 27 above)
Comment (nainkunal933) [reply test/devicecontroller/DefaultDeviceController.test.ts:2418]: Done

File: test/devicecontroller/DefaultDeviceController.test.ts (hunk 29 above)
Comment (ltrung) [reply test/devicecontroller/DefaultDeviceController.test.ts:2502]: How about:
> The current DOMMockBuilder does not set active property of media stream to false if the media stream track is stopped which is not consistent to the browser behavior so we have to manually set the property here.

File: demos/browser/app/meetingV2/meetingV2.ts (hunk 10 above)
Comment (nainkunal933) [reply demos/browser/app/meetingV2/meetingV2.ts:2366]: I have refactored the two methods.

File: README.md (hunk 14 above)
Comment (nainkunal933) [reply README.md:216]: I have removed the quote.

File: demos/browser/app/meetingV2/meetingV2.ts (hunk 24 above)
Comment (nainkunal933) [reply demos/browser/app/meetingV2/meetingV2.ts:751]: I have included the check for unified plan.

File: demos/browser/app/meetingV2/meetingV2.ts [hunk 31]
```diff
@@ -751,6 +753,17 @@ export class DemoMeetingApp
       );
     });
 
+    if(!this.defaultBrowserBehaviour.supportsCanvasCapturedStreamPlayback() || !this.enableUnifiedPlanForChromiumBasedBrowsers)  {
```
Comment (ltrung) [inline demos/browser/app/meetingV2/meetingV2.ts:756]: Can this check move to a private method?

Comment (nainkunal933) [reply demos/browser/app/meetingV2/meetingV2.ts]: Done

File: demos/browser/app/meetingV2/meetingV2.ts [hunk 32]
```diff
@@ -2417,6 +2463,15 @@ export class DemoMeetingApp
     );
   }
 
+  private areVideoFiltersSupported(): boolean {
+    if (this.defaultBrowserBehaviour.supportsCanvasCapturedStreamPlayback() && this.enableUnifiedPlanForChromiumBasedBrowsers)  {
```
Comment (ltrung) [inline demos/browser/app/meetingV2/meetingV2.ts:2467]: nit: you can just do return this.defaultBrowserBehaviour.supportsCanvasCapturedStreamPlayback() && this.enableUnifiedPlanForChromiumBasedBrowsers without the if-else

Comment (nainkunal933) [reply demos/browser/app/meetingV2/meetingV2.ts:2467]: Done

**Review by ltrung — APPROVED** [review]
